src/sentry/snuba/outcomes.py

```diff
@@ -0,0 +1,339 @@
+        if len(raw_fields) == 0:
+            raise InvalidField('Request is missing a "field"')
+
+        self.fields = {}
+        self.aggregations = []
+        self.query: List[Any] = []  # not used but needed for compat with sessions logic
+        start, end, rollup = get_constrained_date_range(query, allow_minute_resolution)
+        self.dataset = _outcomes_dataset(rollup)
+        self.rollup = rollup
+        self.start = start
+        self.end = end
+
+        for key in raw_fields:
+            if key not in COLUMN_MAP:
+                raise InvalidField(f'Invalid field: "{key}"')
+            field = COLUMN_MAP[key]
+            self.aggregations.append(field.aggregation(self.dataset))
+            self.fields[key] = field
+
+        self.groupby = []
+        for key in raw_groupby:
+            if key not in GROUPBY_MAP:
+                raise InvalidField(f'Invalid groupBy: "{key}"')
+            self.groupby.append(GROUPBY_MAP[key])
+
+        if len(query.getlist("category", [])) == 0 and "category" not in raw_groupby:
+            raise InvalidQuery("Query must have category as groupby or filter")
+
+        query_columns = set()
+        for field in self.fields.values():
+            query_columns.update(field.get_snuba_columns(raw_groupby))
+        for groupby in self.groupby:
+            query_columns.update(groupby.get_snuba_columns())
+        self.query_columns = list(query_columns)
+
+        query_groupby = set()
+        for groupby in self.groupby:
+            query_groupby.update(groupby.get_snuba_groupby())
+        self.query_groupby = list(query_groupby)
+
+        self.conditions, self.filter_keys = get_filter(query, params)
+
+
+def run_outcomes_query(query: QueryDefinition) -> Tuple[ResultSet, ResultSet]:
+    result = raw_query(
+        dataset=query.dataset,
+        start=query.start,
+        end=query.end,
+        groupby=query.query_groupby,
+        aggregations=query.aggregations,
+        rollup=query.rollup,
+        filter_keys=query.filter_keys,
+        conditions=query.conditions,
+        selected_columns=query.query_columns,
+        referrer="outcomes.totals",
+        # orderby=query.orderby, TODO: add orderby?
+    )
+    result_timeseries = raw_query(
+        dataset=query.dataset,
+        selected_columns=[TS_COL] + query.query_columns,
+        groupby=[TS_COL] + query.query_groupby,
+        aggregations=query.aggregations,
+        conditions=query.conditions,
+        filter_keys=query.filter_keys,
+        start=query.start,
+        end=query.end,
+        rollup=query.rollup,
+        referrer="outcomes.timeseries",
+    )
+
+    result_totals = _format_rows(result["data"], query)
+    result_timeseries = _format_rows(result_timeseries["data"], query)
+    return result_totals, result_timeseries
+
+
+def _format_rows(rows: ResultSet, query: QueryDefinition) -> ResultSet:
+    category_grouping: Dict[str, Any] = {}
+
+    def _group_row(row: Dict[str, Any]) -> None:
```

Comment:
will definitely want to see performance of this in staging. there are some easier optimizations to be had with this function (we only need to do it on rows category = error, security, default). would be an issue in orgs with large project cardinality. additionally, it could be done in the massaging step so we're not looping through twice.

Reply:
there is also the option of doing the groupby in the snuba query instead of here, which is also an option.